### PR TITLE
stackit: docs and userdata

### PR DESCRIPTION
**Pull request by malt3**
### Context

Some more polishing for STACKIT.

### Proposed change(s)

- move credentials to instance user data
- add all STACKIT availability zones
- explain recovery on STACKIT

### Checklist

- [x] Run the E2E tests that are relevant to this PR's changes
- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: docs/docs/workflows/recovery.md [hunk 1]
```diff
@@ -118,6 +118,37 @@ If this fails due to an unhealthy control plane, you will see log messages simil
 
 This means that you have to recover the node manually.
 
+</tabItem>
+<tabItem value="stackit" label="STACKIT">
+
+First, open the STACKIT portal to view all servers in your project. Select individual control plane nodes `<cluster-name>-<UID>-control-plane-<UID>-<index>` and check that enough members are in a *Running* state.
+
+Second, check the boot logs of these *Servers*. Click on a server name and select **Overview**. Find the **Machine Setup** section and click on **Web console** **Open console**.
```
Comment (burgerdev) [inline docs/docs/workflows/recovery.md:126]:
```suggestion
Second, check the boot logs of these servers. Click on a server name and select **Overview**. Find the **Machine Setup** section and click on **Web console** **Open console**.
```

File: docs/docs/workflows/recovery.md [hunk 2]
```diff
@@ -118,6 +118,37 @@ If this fails due to an unhealthy control plane, you will see log messages simil
...
+
+In the serial console output, search for `Waiting for decryption key`.
+Similar output to the following means your node was restarted and needs to decrypt the [state disk](../architecture/images.md#state-disk):
+
+```json
+{"level":"INFO","ts":"2022-09-08T10:21:53Z","caller":"cmd/main.go:55","msg":"Starting disk-mapper","version":"2.0.0","cloudProvider":"gcp"}
+{"level":"INFO","ts":"2022-09-08T10:21:53Z","logger":"setupManager","caller":"setup/setup.go:72","msg":"Preparing existing state disk"}
+{"level":"INFO","ts":"2022-09-08T10:21:53Z","logger":"rejoinClient","caller":"rejoinclient/client.go:65","msg":"Starting RejoinClient"}
+{"level":"INFO","ts":"2022-09-08T10:21:53Z","logger":"recoveryServer","caller":"recoveryserver/server.go:59","msg":"Starting RecoveryServer"}
+```
```
Comment (burgerdev) [inline docs/docs/workflows/recovery.md:136]: The keyword `Waiting for decryption key` does not appear in the json logs - is that intentional?

File: docs/docs/workflows/recovery.md [hunk 3]
```diff
@@ -118,6 +118,37 @@ If this fails due to an unhealthy control plane, you will see log messages simil
...
+
+The node will then try to connect to the [*JoinService*](../architecture/microservices.md#joinservice) and obtain the decryption key.
```
Comment (burgerdev) [inline docs/docs/workflows/recovery.md:138]: "The node will then try" sounds like the user did something, which makes the node do something, but we did not ask the user to do something. Maybe replace with "If you see something like this further down in the logs ..."?

File: internal/cloud/openstack/imds.go [hunk 4]
```diff
@@ -129,82 +131,89 @@ func (c *imdsClient) role(ctx context.Context) (role.Role, error) {
...
 func (c *imdsClient) timeForUpdate(t time.Time) bool {
 	return time.Since(t) > maxCacheAge
 }
 
-// update updates instance metadata from the azure imds API.
 func (c *imdsClient) update(ctx context.Context) error {
+	if err := c.updateInstanceMetadata(ctx); err != nil {
+		return err
+	}
+	return c.updateUserData(ctx)
```
Comment (burgerdev) [inline internal/cloud/openstack/imds.go:212]: Nit: consider prefixing/wrapping the errors to indicate which data request failed.

**Review by burgerdev — APPROVED** [review]

File: docs/docs/workflows/recovery.md (hunk 2 above)
Comment (malt3) [reply docs/docs/workflows/recovery.md:136]: This is verbatim copied from the other examples for Azure, AWS, and GCP. I would prefer to fix them all together as a follow-up if that’s ok with you.

File: docs/docs/workflows/recovery.md (hunk 3 above)
Comment (malt3) [reply docs/docs/workflows/recovery.md:138]: This is verbatim copied from the other examples for Azure, AWS, and GCP. I would prefer to fix them all together as a follow-up if that’s ok with you.

Comment (burgerdev) [reply docs/docs/workflows/recovery.md:138]: It is.

File: docs/docs/workflows/recovery.md (hunk 2 above)
Comment (burgerdev) [reply docs/docs/workflows/recovery.md:136]: It is.

File: docs/docs/workflows/recovery.md [hunk 5]
```diff
@@ -118,6 +118,37 @@ If this fails due to an unhealthy control plane, you will see log messages simil
 
 This means that you have to recover the node manually.
 
+</tabItem>
+<tabItem value="stackit" label="STACKIT">
+
+First, open the STACKIT portal to view all servers in your project. Select individual control plane nodes `<cluster-name>-<UID>-control-plane-<UID>-<index>` and check that enough members are in a *Running* state.
+
+Second, check the boot logs of these servers. Click on a server name and select **Overview**. Find the **Machine Setup** section and click on **Web console** **Open console**.
```
Comment (thomasten) [inline docs/docs/workflows/recovery.md:126]:
```suggestion
Second, check the boot logs of these servers. Click on a server name and select **Overview**. Find the **Machine Setup** section and click on **Web console** > **Open console**.
```

File: internal/cloud/openstack/imds.go [hunk 6]
```diff
@@ -218,6 +230,20 @@ func (c *imdsClient) update(ctx context.Context) error {
...
 }
 
+func (c *imdsClient) updateUserData(ctx context.Context) error {
+	resp, err := httpGet(ctx, c.client, imdsUserDataURL)
+	if err != nil {
+		return err
+	}
+	var userdataResp userDataResponse
+	if err := json.Unmarshal(resp, &userdataResp); err != nil {
+		return fmt.Errorf("unmarshalling IMDS user_data response %q: %w", string(resp), err)
```
Comment (thomasten) [inline internal/cloud/openstack/imds.go:240]:
```suggestion
		return fmt.Errorf("unmarshalling IMDS user_data response %q: %w", resp, err)
```
nit: superfluous cast (handled by %q)

File: internal/cloud/openstack/imds.go [hunk 7]
```diff
@@ -218,6 +230,20 @@ func (c *imdsClient) update(ctx context.Context) error {
...
+	}
+	c.userDataCache = userdataResp
+	c.cacheTime = time.Now()
```
Comment (thomasten) [inline internal/cloud/openstack/imds.go:243]: nit: should cacheTime be set in update() instead?

**Review by thomasten — APPROVED** [review]

**Pull request merged**